### PR TITLE
Add style checker for beginners

### PR DESCRIPTION
**Pull request by JackMorganNZ**
This pull request adds the first style checker for beginners, including language statistics.

### CODE REVIEW

File: infrastructure/prod-deploy/update-content.sh [hunk 1]
```diff
@@ -16,5 +16,6 @@ source ./codewof/load-prod-envs.sh
 ./dev migrate
 docker-compose exec django /docker_venv/bin/python3 ./manage.py load_user_types
 ./dev load_questions
+./dev load_style_errors
```
Comment (courtneycb) [inline infrastructure/prod-deploy/update-content.sh:19]: Are we not adding this to dev-deploy as well?

File: requirements/base.txt [hunk 2]
```diff
@@ -30,3 +31,10 @@ google-api-python-client==1.7.12
 verto==0.10.0
 python-markdown-math==0.6
 PyYAML==5.3.1
+
+# Style checker
+# Replace flake8 dependency once new release is created
```
Comment (courtneycb) [inline requirements/base.txt:36]: Might need to make a note somewhere else or an issue to do this so we don't forget about it after it's merged

File: setup.cfg [hunk 3]
```diff
@@ -11,14 +11,19 @@ exclude =
     */temp,
     */manage.py,
     codewof/programming/content/en/*/initial.py,
+    */tests/*,
```
Comment (courtneycb) [inline setup.cfg:14]: why?

File: codewof/style/style_checkers/python3_data.py [hunk 4]
```diff
@@ -0,0 +1,1132 @@
+"""Data for Python 3 style errors.
+
+The following fields are rendered as HTML.
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/style/style_checkers/python3_data.py:3]: :sob:

File: codewof/style/style_checkers/python3_data.py [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+    },
+    "W604": {
+    "original_message": "backticks are deprecated, use ‘repr()’",
+    "title": "Backticks have been removed in Python 3.",
+    "solution": "Use the built-in function repr() instead.",
+    "explanation": ""
+    },
+    "W605": {
+    "original_message": "invalid escape sequence ‘x’",
+    "title": "",
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/style/style_checkers/python3_data.py]: Titles have now been added.

File: codewof/templates/style/language-statistics.html [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    {% breadcrumb "Statistics" "style:language_statistics" language.slug %}
+{% endblock breadcrumbs %}
+
+{% block page_heading %}
+    {% include language_header %}
+    <h2 class="mt-3">
+        Statistics
+    </h2>
+    <p>The following table shows the occurence counts of style issues submitted to the {{ language.name }} style checker.</p>
+    <p>The text <code>{article} {character_description}</code> will mean the style issue involves one of the following characters: <code>{{ characters|join:"" }}</code></p>
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/templates/style/language-statistics.html:21]: Improved with new `simplify_error_template` tag.

File: codewof/style/style_checkers/python3_data.py [hunk 7]
```diff
@@ -0,0 +1,952 @@
...
+        "title_templated": False,
+        "title": "Line is missing a blank line between the methods of a class.",
+        "solution": "Add a blank line in between your class methods.",
+        "explanation": ""
+    },
+    "E302": {
+        "original_message": "expected 2 blank lines, found 0",
+        "title_templated": False,
+        "title": "Two blank lines are expected between functions and classes.",
+        "solution": "Ensure there are two blank lines between functions and classes.",
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/style/style_checkers/python3_data.py:335]: Updated text.

File: codewof/style/style_checkers/python3_data.py [hunk 8]
```diff
@@ -0,0 +1,1132 @@
+"""Data for Python 3 style errors.
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/style/style_checkers/python3_data.py:1]: No, that error is just a straight syntax error which I've written a nicer message for.

File: codewof/style/style_checkers/python3.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+LINE_RE = re.compile(r':(?P<line>\d+):(?P<character>\d+): (?P<error_code>\w\d+) (?P<error_message>.*)$')
+CHARACTER_RE = re.compile(r'\'(?P<character>.*)\'')
+TEMP_FILE_ROOT = settings.STYLE_CHECKER_TEMP_FILES_ROOT
+TEMP_FILE_EXT = '.py'
+# Create folder if it does not exist
+Path(TEMP_FILE_ROOT).mkdir(parents=True, exist_ok=True)
+PYTHON3_DETAILS = get_language_info('python3')
+
+
+def python3_style_check(code):
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/style/style_checkers/python3.py:27]: Yeah, will look into at a later time.

File: codewof/general/management/commands/sampledata.py [hunk 10]
```diff
@@ -44,7 +44,7 @@ def handle(self, *args, **options):
 
         management.call_command('load_user_types')
         print(LOG_HEADER.format('Create sample users'))
-        User = get_user_model()
+        User = get_user_model()  # noqa N806
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/general/management/commands/sampledata.py:47]: 'variable in function should be lowercase' but we are importing a class so keeping the same case.

File: infrastructure/prod-deploy/update-content.sh (hunk 1 above)
Comment (JackMorganNZ) [reply infrastructure/prod-deploy/update-content.sh:19]: This is included in the `sampledata` command.

File: codewof/style/style_checkers/python3_data.py [hunk 11]
```diff
@@ -0,0 +1,1132 @@
...
+    },
+    "E131": {
+        "original_message": "continuation line unaligned for hanging indent",
+        "title_templated": False,
+        "title": "Line is not aligned correctly for a hanging indent.",
+        "solution": "Add or remove indentation so that the lines are aligned with each other.",
+        "explanation": ""
+    },
+    # E133 ignored by default
+    # "E133": {
```
[the remark this answers is not in the document]
Comment (JackMorganNZ) [reply codewof/style/style_checkers/python3_data.py:139]: Ignore a few, but all are now specifically ignored in the `flake8.ini` file.

File: codewof/config/templatetags/simplify_error_template.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+"""Module for the custom simplify_error_template template tag."""
+
+from django import template
+from django.utils.safestring import mark_safe
+
+register = template.Library()
+
+SEARCH_TEXT = '{article} {character_description}'
+REPLACE_TEXT = 'a {character}'
```
Comment (courtneycb) [inline codewof/config/templatetags/simplify_error_template.py:9]: baziiing how good is this :squid-dab:

Comment (JackMorganNZ) [reply codewof/config/templatetags/simplify_error_template.py:9]: :tada:

**Pull request merged**